libibumad/umad.c

```diff
@@ -1123,3 +1211,73 @@ void umad_dump(void *umad)
 	       mad->agent_id, mad->status, mad->timeout_ms);
 	umad_addr_dump(&mad->addr);
 }
+
+struct umad_device_node *umad_get_ca_device_list(void)
+{
+	DIR *dir;
+	struct dirent *entry;
+	struct umad_device_node *head = NULL;
+	struct umad_device_node *tail;
+	struct umad_device_node *node;
+	char *ca_name;
+	size_t cas_num = 0;
+	size_t d_name_size, node_size;
+
+	dir = opendir(SYS_INFINIBAND);
+	if (!dir) {
+		if (errno == ENOENT)
+			errno = 0;
+		goto exit;
+	}
+
+	while ((entry = readdir(dir))) {
+		if ((strcmp(entry->d_name, ".") == 0) ||
+		    (strcmp(entry->d_name, "..") == 0))
+			continue;
+
+		if (!is_ib_type(entry->d_name))
+			continue;
+
+		d_name_size = strlen(entry->d_name) + 1;
+		node_size = sizeof(struct umad_device_node);
+		node = calloc(1, node_size + d_name_size);
+		if (!node) {
+			errno = ENOMEM;
+			umad_free_ca_device_list(head);
+			head = NULL;
+			goto exit;
+		}
+
+		if (!head)
+			head = node;
+		else
+			tail->next = node;
+		tail = node;
+
+		ca_name = (char *)node + node_size;
+		strncpy(ca_name, entry->d_name, d_name_size);
+		node->ca_name = ca_name;
+
+		cas_num++;
+	}
+
+	DEBUG("return %zu cas", cas_num);
+	if (!cas_num)
```

Comment:
cas_num isn't really used, just drop it and always set errno to 0 on a success exit.